compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -0,0 +1,189 @@
+  private def paramsOnlyP(trigger: Symbol => Boolean)(implicit ctx: Context): MethodP = {
+    MethodP(trigger, nnParams = Seq.empty, nnRes = true)
+  }
+
+  /** Nullifies a Java type by adding `| JavaNull` in the relevant places. */
+  private def nullifyType(tpe: Type)(implicit ctx: Context): Type = {
+    val nullMap = new JavaNullMap(alreadyNullable = false)
+    nullMap(tpe)
+  }
+
+  /** A type map that adds `| JavaNull`.
+   *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
+   *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
+   *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
+   */
+  private class JavaNullMap(alreadyNullable: Boolean)(implicit ctx: Context) extends TypeMap {
+    /** Should we nullify `tp` at the outermost level? */
+    def needsTopLevelNull(tp: Type): Boolean = {
+      !alreadyNullable && (tp match {
+        case tp: TypeRef =>
+          // We don't modify value types because they're non-nullable even in Java.
+          // We don't modify `Any` or `RefEq` because they're already nullable.
+          !tp.symbol.isValueClass && !tp.isRef(defn.AnyClass) && !tp.isRef(defn.RefEqClass)
+        case _ => true
+      })
+    }
+
+    /** Should we nullify the type arguments to the given generic `tp`?
+     *  We only nullify the inside of Scala-defined generics.
+     *  This is because Java classes are _all_ nullified, so both `java.util.List[String]` and
+     *  `java.util.List[String|Null]` contain nullable elements.
+     */
+    def needsNullArgs(tp: AppliedType): Boolean = {
+      val AppliedType(tycons, _) = tp
+      tycons.widenDealias match {
+        case tp: TypeRef if !tp.symbol.is(JavaDefined) => true
+        case _ => false
+      }
+    }
+
+    override def apply(tp: Type): Type = {
+      tp match {
+        case tp: LambdaType => mapOver(tp)
+        case tp: TypeAlias => mapOver(tp)
+        case tp@AndType(tp1, tp2) =>
+          // nullify(A & B) = (nullify(A) & nullify(B)) | JavaNull, but take care not to add
+          // duplicate `JavaNull`s at the outermost level inside `A` and `B`.
+          val newMap = new JavaNullMap(alreadyNullable = true)
+          derivedAndType(tp, newMap(tp1), newMap(tp2)).toJavaNullableUnion
+        case tp@OrType(tp1, tp2) if !tp.isJavaNullableUnion =>
+          val newMap = new JavaNullMap(alreadyNullable = true)
+          derivedOrType(tp, newMap(tp1), newMap(tp2)).toJavaNullableUnion
+        case tp: TypeRef if needsTopLevelNull(tp) => tp.toJavaNullableUnion
+        case tp: TypeParamRef if needsTopLevelNull(tp) => tp.toJavaNullableUnion
+        case appTp@AppliedType(tycons, targs) =>
+          val targs2 = if (needsNullArgs(appTp)) targs map this else targs
+          derivedAppliedType(appTp, tycons, targs2).toJavaNullableUnion
+        case _ => tp
+      }
```

Comment:
-  (spaces) Done.
-  (common cases to the top) Done.
- (avoid creating JavaNullMap). Done.
- (remove first two cases). I can't: otherwise I turn

[T <: Comparable[_ >: LazyRef(T)]](x$0: java.util.Collection[_ <: T]): T 
into
[T >: Nothing | JavaNull <: Comparable[_ >: LazyRef(T)] | JavaNull]
  (x$0: java.util.Collection[_ <: T] | JavaNull): T | JavaNull

I think in this specific case what happens is we would `mapOver` the `TypeBounds` for each of the 
type arguments of a `PolyType`. We could special case type bounds to do nothing, so we don't recurse
inside of them, but then the current opt-in seems to be safer than the opt-out approach you're suggesting.

Reply:
Thanks for the explanation.